### PR TITLE
Implement `hidden=until-found` HTML attribute and `beforematch` event

### PR DESCRIPTION
**Pull request by nt1m**
#### 4f12f2494d1ca0a52e7bee8a94bbc94194faf48a
<pre>
Implement `hidden=until-found` HTML attribute and `beforematch` event
<a href="https://bugs.webkit.org/show_bug.cgi?id=238266">https://bugs.webkit.org/show_bug.cgi?id=238266</a>
<a href="https://rdar.apple.com/90706206">rdar://90706206</a>

Reviewed by Darin Adler.

Spec: <a href="https://html.spec.whatwg.org/#the-hidden-attribute">https://html.spec.whatwg.org/#the-hidden-attribute</a>

* LayoutTests/TestExpectations:
* LayoutTests/editing/find/cocoa/find-in-page-in-hidden-until-found-expected.txt: Added.
* LayoutTests/editing/find/cocoa/find-in-page-in-hidden-until-found.html: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-hidden-until-found-expected.txt: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-hidden-until-found.html: Added.
Find in page tests.

* LayoutTests/imported/w3c/web-platform-tests/html/editing/the-hidden-attribute/beforematch-element-fragment-navigation-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/editing/the-hidden-attribute/hidden-ua-stylesheet-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/trusted-types/trusted-types-event-handlers-expected.txt:
Rebaseline test passes.

* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.json:
* Source/WebCore/dom/FindRevealAlgorithms.cpp:
(WebCore::revealHiddenUntilFoundAncestors): Implement <a href="https://html.spec.whatwg.org/#ancestor-hidden-until-found-revealing-algorithm">https://html.spec.whatwg.org/#ancestor-hidden-until-found-revealing-algorithm</a>
(WebCore::revealClosedDetailsAndHiddenUntilFoundAncestors):
* Source/WebCore/dom/FindRevealAlgorithms.h:

* Source/WebCore/dom/GlobalEventHandlers.idl:
* Source/WebCore/html/HTMLAttributeNames.in:
Add beforematch event.

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::collectPresentationalHintsForAttribute): `hidden=until-found` uses `content-visibility: hidden`
(WebCore::HTMLElement::isHiddenUntilFound const): Helper method to avoid duplication
* Source/WebCore/html/HTMLElement.h:

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragmentInternal): Follow <a href="https://html.spec.whatwg.org/#scrolling-to-a-fragment">https://html.spec.whatwg.org/#scrolling-to-a-fragment</a>

* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const): Make sure autoRevealsWhenFound bit is set so that TextIterator doesn&apos;t skip these subtrees.

* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::didFindString):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::decorateTextRangeWithStyle):
* Source/WebKit/WebProcess/WebPage/ios/FindControllerIOS.mm:
(WebKit::FindController::didFindString):

3 different codepaths for find-in-page on different configurations.

Canonical link: <a href="https://commits.webkit.org/297073@main">https://commits.webkit.org/297073@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/7cffa86b1a5dae32eaa89d25ee98df688efe38c7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110501 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/30160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20593 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/116526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/60768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112464 "Passed tests") | [❌ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30839 "Failed to checkout and rebase branch from PR 47284") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38748 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/116526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/60768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113449 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/30839 "Failed to checkout and rebase branch from PR 47284") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/116526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/30839 "Failed to checkout and rebase branch from PR 47284") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK1-Tests-EWS "Waiting to run tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/60322 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/30839 "Failed to checkout and rebase branch from PR 47284") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/17695 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/119317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/37541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/27860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/119317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/37915 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/95767 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/119317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/33514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17818 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/37436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/42907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/37098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/40438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/38807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/WebCore/page/LocalFrameView.cpp [hunk 1]
```diff
@@ -2743,13 +2743,17 @@ bool LocalFrameView::scrollToFragmentInternal(StringView fragmentIdentifier)
         return false;
     }
 
+    if (anchorElement) {
+        revealClosedDetailsAncestors(*anchorElement);
+        revealHiddenUntilFoundAncestors(*anchorElement);
```
Comment (nt1m) [inline Source/WebCore/page/LocalFrameView.cpp]: Possible followup:

Perhaps both of these methods should be moved to a new `Source/WebCore/dom/FindReveal.h` file (I'm bad at names, please suggest alternatives!) instead of being spread over HTMLElement and HTMLDetailsElement?

That new file could also contain a helper method that calls both to avoid the duplication.

Opinions welcome!

File: Source/WebCore/page/LocalFrameView.cpp [hunk 2]
```diff
@@ -2743,13 +2743,17 @@ bool LocalFrameView::scrollToFragmentInternal(StringView fragmentIdentifier)
...
+    }
+
     RefPtr<ContainerNode> scrollPositionAnchor = anchorElement;
     if (!scrollPositionAnchor)
         scrollPositionAnchor = m_frame->document();
     maintainScrollPositionAtAnchor(scrollPositionAnchor.get());
-    
+
     if (anchorElement) {
-        revealClosedDetailsAncestors(*anchorElement);
```
Comment (nt1m) [inline Source/WebCore/page/LocalFrameView.cpp:2780]: (This is moved up to comply with the spec, which says to do scrolling should happen _after_ the two reveal algorithms are invoked. This is exercised by many of the beforematch-element-fragment-navigation.html subtests.) 

Some tests could probably be written for the details element auto-expansion as a followup too.

File: Source/WebCore/html/HTMLElement.cpp [hunk 3]
```diff
@@ -1367,6 +1377,23 @@ void HTMLElement::setPopover(const AtomString& value)
...
+{
+    if (!node.document().settings().hiddenUntilFoundEnabled())
+        return;
+
+    Ref currentNode = node;
+    while (RefPtr parentElement = currentNode->parentElementInComposedTree()) {
+        RefPtr currentElement = dynamicDowncast<HTMLElement>(currentNode);
+        if (currentElement && currentElement->isHiddenUntilFound()) {
+            currentElement->dispatchEvent(Event::create(eventNames().beforematchEvent, Event::CanBubble::Yes, Event::IsCancelable::No));
+            currentElement->setHidden(std::nullopt);
```
Comment (annevk) [inline Source/WebCore/html/HTMLElement.cpp]:
```suggestion
            currentElement->setHidden({ });
```

File: Source/WebCore/page/LocalFrameView.cpp (hunk 1 above)
Comment (annevk) [reply Source/WebCore/page/LocalFrameView.cpp]: Doesn't this current structure mean we perform two ancestor walks? I'm pretty sure that won't be considered acceptable?

Comment (nt1m) [reply Source/WebCore/page/LocalFrameView.cpp]: I can test the absence of the `autoExpandsWhenFound` style bit (which will be the majority of nodes fwiw) to remove both ancestor walks.

I'd rather keep those as two separate walks for readability, this is closer to the spec and also how other browsers do it. Given this is only called in very specific situations, I don't think it's worth optimizing it into a single walk. The style bit optimization should be enough.

Comment (annevk) [reply Source/WebCore/page/LocalFrameView.cpp]: Isn't this called whenever you do find-in-page (well not this one, but where they are coupled elsewhere)? That is quite frequent for me at least and I'm not too impressed with the current performance either.

Comment (nt1m) [reply Source/WebCore/page/LocalFrameView.cpp]: Yes though the style bit optimization that I'll add will only make sure they're not called on non-hidden=until-found/details subtrees

(Fwiw, I'm trying to get the correctness down first before adding the optimizations I have in mind. Right now this code won't run when the preference is off, so that seems OK?)

Comment (nt1m) [reply Source/WebCore/page/LocalFrameView.cpp]: @annevk Collapsing into a single ancestor walk would actually be web observable if you have hidden=until-found used on a element slotted into a details element. The order of the details uncollapsing and the beforematch events would be different

**Comment by nt1m** [comment]
This is ready for review!

File: Source/WebCore/dom/FindRevealAlgorithms.cpp [hunk 4]
```diff
@@ -53,4 +54,33 @@ void revealClosedDetailsAncestors(Node& node)
...
+
+    Ref currentNode = node;
+    while (RefPtr parentElement = currentNode->parentElementInComposedTree()) {
+        RefPtr currentElement = dynamicDowncast<HTMLElement>(currentNode);
+        if (currentElement && currentElement->isHiddenUntilFound()) {
+            currentElement->dispatchEvent(Event::create(eventNames().beforematchEvent, Event::CanBubble::Yes, Event::IsCancelable::No));
+            currentElement->setHidden({ });
+        }
+        currentNode = *parentElement;
+    }
```
Comment (darinadler) [inline Source/WebCore/dom/FindRevealAlgorithms.cpp]: We’d normally want to use a `for` loop instead of a `while` loop if possible so we can see the structure of the iteration. I prefer single word variable names over multi-word ones unless there is some compelling reason we have to use multi-word ones; local variable names don’t need to express a lot of context, and words are easy to read.

Can we use `composedTreeAncestors`?
```suggestion
    for (Ref ancestor : composedTreeAncestors()) {
        RefPtr element = dynamicDowncast<HTMLElement>(ancestor);
        if (element && element->isHiddenUntilFound()) {
            element->dispatchEvent(Event::create(eventNames().beforematchEvent, Event::CanBubble::Yes, Event::IsCancelable::No));
            element->setHidden({ });
        }
    }
```
Or is iterating like this a problem when the event mutates the tree? Do we have test cases that cover the correct interoperable behavior when the element is changed by the `beforematch` event handling code?

My suggestion won’t solve this potential problem: The loop over composed tree ancestors could be an infinite loop if `beforematch` event handlers keep changing the ancestry of elements back and forth, alternately putting one element inside the other and setting `hidden` back to `until-found`. I suggest we design the algorithms so they can’t create infinite loops and add stress test cases for WPT that would fail if the algorithm was wrong.

**Review by darinadler — APPROVED** [review]

File: Source/WebCore/dom/FindRevealAlgorithms.cpp (hunk 4 above)
Comment (nt1m) [reply Source/WebCore/dom/FindRevealAlgorithms.cpp]: I'll switch to a for loop, no reason not to here :)

I've filed https://github.com/whatwg/html/issues/11436 to address the infinite loop issue.

Comment (nt1m) [reply Source/WebCore/dom/FindRevealAlgorithms.cpp]: Ah the `composedTreeAncestors` doesn't quite work here, since I need the lineage, but I can probably still use a for loop

Comment (darinadler) [reply Source/WebCore/dom/FindRevealAlgorithms.cpp]: The `for` loop you landed is different from the `while` loop in that the parent operation is computed after dispatching the event and calling `setHidden`; in the `while` loop it was done before. That is an observable difference. I think we should have stress tests for such things in WPT; these bizarre edge cases can still end up causing website incompatibilities!

Comment (nt1m) [reply Source/WebCore/dom/FindRevealAlgorithms.cpp]: Great catch, I'll wait for the outcome of https://github.com/whatwg/html/issues/11436 and adjust algorithms / write WPTs accordingly!
